lingua_franca/internal.py

```diff
@@ -0,0 +1,587 @@
+
+        @localized_function(run_own_code_on=[UnsupportedLanguageError])
+        def nice_number(number, lang=None, speech=True, denominators=None):
+
+    Here, nice_number() itself will be executed in the event that the localizer
+    raises an UnsupportedLanguageError.
+
+    Arguments:
+        run_own_code_on(list(type), optional)
+            A list of Error types (ValueError, NotImplementedError, etc)
+            which, if they are raised, will trigger the wrapped function's
+            own code.
+
+            If this argument is omitted, the function itself will never
+            be run. Calls to the wrapped function will be passed to the
+            appropriate, localized function.
+
+
+    """
+    # Make sure everything in run_own_code_on is an Error or None
+    def is_error_type(_type):
+        _instance = _type()
+        rval = isinstance(_instance, BaseException) if _instance else True
+        del _instance
+        return rval
+    if not isinstance(run_own_code_on, list):
+        run_own_code_on = list(run_own_code_on)
+    if run_own_code_on != [None]:
+        if not all((is_error_type(e) for e in run_own_code_on)):
+            raise ValueError("@localized_function(run_own_code_on=<>) expected an "
+                             "Error type, or a list of Error types. Instead, it "
+                             "received this value:\n" + str(run_own_code_on))
+
+    # Begin wrapper
+    def localized_function_decorator(func):
+        @wraps(func)
+        def call_localized_function(*args, **kwargs):
+            try:
+                func_signature = signature(func)
+                func_params = list(func_signature.parameters)
+                lang_param_index = func_params.index('lang')
+
+                # Momentarily assume we're not passing a lang code
+                lang_code = get_default_lang()
+                full_lang_code = __active_lang_code
+
+                # Check if we're passing a lang as a kwarg
+                if 'lang' in kwargs.keys():
+                    lang_code = kwargs['lang']
+
+                # Check if we're passing a lang as a positional arg
+                elif lang_param_index < len(args):
+                    if args[lang_param_index] in (_SUPPORTED_LANGUAGES,
+                                                  _SUPPORTED_FULL_LOCALIZATIONS):
+                        lang_code = args[lang_param_index]
+                if not lang_code:
+                    raise ModuleNotFoundError("No language module loaded.")
+
+                if lang_code not in _SUPPORTED_LANGUAGES:
+                    try:
+                        lang_code = get_primary_lang_code(lang_code)
+                    except ValueError:
+                        _raise_unsupported_language(lang_code)
+                    if lang_code not in _SUPPORTED_LANGUAGES:
+                        _raise_unsupported_language(lang_code)
+                    full_lang_code = lang_code
```

Comment:
This nested, repeated check and raise might warrant explanation. It is hypothetically possible for a developer to register a new language everywhere *but* _SUPPORTED_LANGUAGES. In that case, `get_primary_lang_code()` would return the author's intended primary code, allowing all of LF's *internal* functions to work, but preventing the localizer from finding it.

The second check will prevent their tests from passing in that case.

Reply:
Personally I'm also a fan of extracting complicated blocks out to their own function leaving behind a `_get_supported_lang_code()` or something. Even if it's only used once I find the improved readability worth it. 

I actually wouldn't be opposed to large primary functions being made up entirely of helper function calls but I think I'm in the minority there.